Instructions for adding a native external library dependency (#105045)
dotnet/runtime: src/native/external/README.md

# Native external libraries

> This folder contains the source code of third party native libraries that are used as dependencies for Runtime.

Reviewer: ```suggestion
This folder contains the source code of vendored third party native libraries that are used as dependencies for dotnet/runtime.
```

Author: Pardon my ignorance: What does vendored refer to? Why would all these libraries be considered as such?

Reviewer: "Vendored in" libraries are libraries we include as source in our repository. https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Strategy-For-External-Source.md has the definition.

Other dependencies (like llvm-project), we maintain as a fork. Others we depend on system packages (openssl, our compiler toolchain). In the future, we may download some (test-only) dependencies that aren't needed in source-build scenarios and would only bloat the repo. This README only applies to vendored dependencies.

### Folder structure

Each native library is roughly structured the following way:

```
library-folder/
    library-source-code-files
library.cmake
library-version.txt
cgmanifest.json
```

- `library-folder/` is where the native source code lives.
- `library.cmake` is the main cmake file we add to build this library from anywhere in this repo.

> - `library-version.txt` contains all the detailed release information for this library.

Reviewer 2: ... and information about any local patches applied to the library.

- `cgmanifest.json` is the official file that describes the source code provenance for each one of the external libraries we consume.

> ### How to add or update a new library

Reviewer: ```suggestion
### How to add or update a vendored library
```

```json
"Registrations": [
  {
    "Component": {
      "Type": "git",
      "Git": {
        "RepositoryUrl": "https://github.com/org/repo",
        "CommitHash": "<commit from which the source code was obtained>"
      }
    }
  }
]
```

3. Open the [THIRD-PARTY-NOTICES.txt](https://github.com/dotnet/runtime/blob/main/THIRD-PARTY-NOTICES.TXT) file in the root folder of the runtime repo. Add or update the copy of the license of this library, making sure the license copy has a header like this:

```
License notice for <library name>
-----------------------
```

4. Add or edit the `<library>-version.txt` file under `src/native/external`. This file should contain:
  - The release version of the downloaded release
  - The hash commit of the downloaded release
  - The URL to the tag of the downloaded release in the format `https://github.com/org/repo/releases/tag/<version_number>`
  - An optional section describing any additional information like:
    - Manual modifications we need to make after copying the source code. For example, deleting unnecessary files or trimming unnecessary code.
    - Important notes about the release, like security fixes.

> 5. Make any necessary cmake changes to properly consume the source code.

Reviewer 2: It is not unusual to make changes in the library sources as part of this step. The information about these changes should be included in ...-version.txt file. The same changes should be proposed for the library upstream so that they are not necessary during the next library update.

6. Submit a PR tagging the area owners as well as the @dotnet/runtime-infrastructure team.

7. Find ways to get notified about new releases for the external dependency. For example, if the source code comes from a GitHub repo, you can subscribe to new releases:
   - Go to the main org/repo page
   - Click on Watch
   - Click on Custom
   - Select the "Releases" and "Security alerts" checkboxes
   - Click on "Apply"

> 8. Consult with the .NET Security experts to make sure we meet all of Microsoft's Open Source guidance, especially regarding security updates and timeline expectations.

Reviewer 3: This ought to happen _before_ a new dependency is added, not at the very end :)

Reviewer 3: 9? (For internal users) Validate that CG detects the dependency.

# Native external libraries

This folder contains the source code of vendored third party native libraries that are used as dependencies for dotnet/runtime.

> We call "vendored libraries" those from which we keep a copy of their source code in this repository, and those copies are built alongside the rest of our repository. The main document describing vendored libraries can be found here: https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Strategy-For-External-Source.md

Reviewer 2: ```suggestion
Vendored libraries are copies of third-party dependencies that are included in the project repository. These copies are built alongside the rest of the repository. The document describing our approach for vendored libraries can be found here: https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Strategy-For-External-Source.md
```
We have not invented the terms "vendoring" or "vendored libraries". It is a standard industry term, so we should not say "We call ...". Everybody calls it like that.
